subworkflows/nf-core/bam_variant_calling_haplotypecaller/main.nf

```diff
@@ -0,0 +1,114 @@
+
+workflow BAM_VARIANT_CALLING_HAPLOTYPECALLER {
+
+    take:
+    ch_bam              // channel: [mandatory] [meta, bam, bai, bed]
+    ch_fasta            // channel: [mandatory] [fasta]
+    ch_fasta_fai        // channel: [mandatory] [fasta_fai]
+    ch_dict             // channel: [mandatory] [dict]
+    ch_dragstr_models   // channel: [optional]  [meta, dragstr_model]
+    ch_dbsnp            // channel: [optional]  [dbsnp]
+    ch_dbsnp_tbi        // channel: [optional]  [dbsnp_tbi]
+
+    strategy            // string:  [mandatory] => The strategy used to run the subworkflow ["scatter","no_scatter"]
+    scatter_count       // integer: [optional]  => The amount of scattered BEDs should be created
+
+    main:
+
+    ch_versions = Channel.empty()
+
+    //
+    // Optional: Scatter the BED files
+    //
+    if (strategy == "scatter" && scatter_count > 1){
+
+        ch_beds = ch_bam.map({ meta, bam, bai, bed -> [ meta, bed ]})
+
+        BEDTOOLS_SPLIT (
+            ch_beds,
+            scatter_count
+        )
+        ch_versions = ch_versions.mix(BEDTOOLS_SPLIT.out.versions.first())
+
+        ch_scattered_bams = ch_bam.combine(BEDTOOLS_SPLIT.out.beds.transpose(), by:0)
+                                .map({ meta, bam, bai, full_bed, scattered_bed ->
+                                    new_meta = meta.clone()
+                                    new_meta.id = scattered_bed.baseName
+                                    new_meta.sample = meta.id
+                                    [ new_meta, bam, bai, scattered_bed ]
+                                })
+
+    } else {
+        ch_scattered_bams = ch_bam
+    }
+
+    //
+    // Call variants using HaplotypeCaller
+    //
+
+    if (ch_dragstr_models){
+        ch_bam_dragstr = ch_scattered_bams.combine(ch_dragstr_models, by: 0)
+    } else {
+        ch_bam_dragstr = ch_scattered_bams
+    }
+
+    ch_haplotypecaller_input = ch_bam_dragstr.map({ meta, bam, bai, bed, dragstr_model=[] ->
+                                                    [ meta, bam, bai, bed, dragstr_model ]
+                                                })
+
+    GATK4_HAPLOTYPECALLER (
+        ch_haplotypecaller_input,
+        ch_fasta,
+        ch_fasta_fai,
+        ch_dict,
+        ch_dbsnp,
+        ch_dbsnp_tbi
+    )
+    ch_versions = ch_versions.mix(GATK4_HAPLOTYPECALLER.out.versions.first())
+
+    //
+    // Optional: Merge scattered VCFs
+    //
+
+    if (strategy == "scatter" && scatter_count > 1){
+
+        ch_merge_input = GATK4_HAPLOTYPECALLER.out.vcf
+                            .combine(GATK4_HAPLOTYPECALLER.out.tbi, by:0)
+                            .map({ meta, vcf, tbi ->
+                                new_meta = [:]
+                                new_meta.id = meta.sample
+                                [ new_meta, vcf, tbi ]
+                            })
+                            .groupTuple()
```

Comment:
Here we definitely need a group key magic or the merging will only happen after all haplotypecaller steps for all samples are run through making this a waiting bottleneck and unefficient

Reply:
That's interesting! I didn't think that was possible. I'll investigate 😎

Reply:
See here for example: https://github.com/nf-core/sarek/blob/bcd7bf9cb98cddec27bb54fb47ee122c09388c02/subworkflows/nf-core/variantcalling/haplotypecaller/main.nf#L91 (and it is all over the sarek code base). In your case it might be slightly simpler to implement since you "know" the number of "scatters". In our case we have to count them in an earlier step in the pipeline, because they are changing depending on the bedfile.

Reply:
Allright, thanks! Should the subworkflow contain the code for dependent scatter counts? I don't know if it will make the subworkflow too bloated. It does seem like something that could improve the efficiency though.

Reply:
since you have the fixed number, you can just supply that number as the second part in the groupkey argument. (or in this case might actually be able to use groupTuple and just set `size:scatter` . (best practice anyways to always set size for groupTuple. Otherwise the operator doesn't know when one group is "full" and when to advance.

Reply:
Thanks for this! It really opened up a lot of possibilities 🥳